Modules/mathmodule.c

```diff
@@ -2016,6 +2016,71 @@ math_factorial(PyObject *module, PyObject *arg)
 }
 
 
+/*[clinic input]
+math.lcm
+
+    x as a: object
+    y as b: object
+    /
+least common multiple of x and y
+[clinic start generated code]*/
+
+static PyObject *
```

Comment:
Okay, let's dig into the refcounting a bit more. I'll use this review to annotate the code line-by-line, for the first part of the function.

General principles:

- any references that this function owns should eventually be freed, either via a DECREF, or by returning the referenced value to the caller (which then receives ownership of that reference)
- but keep ownership of references until you no longer need it: don't do a `Py_DECREF(a)` if you still need to use `a` later in the code
- most Python C-API function calls (`PyNumber_Index`, `_PyLong_GCD`, ...) will create and return a new reference, that this function then owns and is reponsible for disposing of .

Let's see how this plays out, step by step, below.